lib/widgets/app.dart

```diff
@@ -152,18 +152,56 @@ class _ZulipAppState extends State<ZulipApp> with WidgetsBindingObserver {
     return super.didPushRouteInformation(routeInformation);
   }
 
-  Future<void> _handleInitialRoute() async {
-    final initialRouteUrl = Uri.parse(WidgetsBinding.instance.platformDispatcher.defaultRouteName);
-    if (initialRouteUrl case Uri(scheme: 'zulip', host: 'notification')) {
-      await NotificationDisplayManager.navigateForNotification(initialRouteUrl);
+  InitialRouteListFactory _handleGenerateInitialRoutes(BuildContext context) {
+    final globalStore = GlobalStoreWidget.of(context);
+
+    void showNotificationErrorDialog() async {
+      final navigator = await ZulipApp.navigator;
+      final navigatorContext = navigator.context;
+      assert(navigatorContext.mounted);
+      // TODO(linter): this is impossible as there's no actual async gap, but
+      //   the use_build_context_synchronously lint doesn't see that.
+      if (!navigatorContext.mounted) return;
+
+      final zulipLocalizations = ZulipLocalizations.of(navigatorContext);
+      showErrorDialog(context: navigatorContext,
+        title: zulipLocalizations.errorNotificationOpenTitle,
+        message: zulipLocalizations.errorNotificationOpenAccountMissing);
     }
+
+    return (String initialRoute) {
+      final initialRouteUrl = Uri.parse(initialRoute);
+      if (initialRouteUrl case Uri(scheme: 'zulip', host: 'notification')) {
+        final notificationResult = NotificationDisplayManager.routeForNotification(
+          globalStore: globalStore,
+          url: initialRouteUrl);
+
+        if (notificationResult != null) {
+          return [
+            HomePage.buildRoute(accountId: notificationResult.accountId),
+            notificationResult.route,
+          ];
```

Comment:
It's a bit unsatisfying to have `routeForNotification` have to return two values like this, when the route already includes the account ID.

I think we can simplify to make it look like this:
```suggestion
          return [
            HomePage.buildRoute(accountId: route.accountId),
            route,
          ];
```

with a small refactor to make the account ID more transparent on the route.

I'll push a small NFC onto your PR branch that makes that refactor. Then you can move it to the front of the branch and make use of it.